Fixed issue #13102 with could_extract_minus_sign

Dirivian opened the pull request:
Put an if statement in could_extract_minus_sign to deal with 
when `expr=-expr`

This is done to fix the bug raised in #13102

smichr commented:
This  is not the desired solution. For expressions like `e = zoo*(-1-x)` we should be able to extract a minus sign to give `zoo*(1 + x)` so the answer should be True but `e == -e`. I think you might to do something like:

```python
if not self.args and self == -self:
    return False
sifted = sift(self.args, lambda i: i == -i)
self = self.func(*sifted[False])
negative_self = -self  # the current first line
...
```
```
Alternatively, negation could be patched by not applying negation to the first arg that changes in expr, so `-(zoo*(1-x)) -> zoo*(x - 1)`.

asmeurer commented:
@smichr I don't think that's helpful because `-zoo*(-1-x)` isn't going to give `zoo*(1 + x)`. It will just give `zoo*(-1-x)`. The functions that use `could_extract_minus_sign` assume that multiplying by a minus sign inverts it. `zoo` messes that up, and I think the only correct solution is to return False as is done here.

asmeurer commented:
@Dirivian please add a test. You can test that both `zoo*(x - 1)` and `zoo*(1 - x)` give False (you can put it with the other `could_extract_minus_sign` tests in `sympy/core/tests/test_expr.py`).

asmeurer commented:
> Alternatively, negation could be patched by not applying negation to the first arg that changes in expr, so -(zoo*(1-x)) -> zoo*(x - 1).

That's a separate thing, and could be a good idea. Honestly these zoo expressions have little meaning anyway. I think this patch should be there regardless, though, as a base guard against the assumptions of the function.

Dirivian commented:
Yeah, I guess the first step is to make sure functions are okay using `zoo` and `could_extract_minus_sign`

And then, when `-(zoo*(.......))` gets resolved, one of the tests should be removed.

Although since a minus sign makes no sense for a complex infinity, could that change cause problems when `zoo` is used in a weird context ?

smichr commented:
How about modifying the docstring:

```
>>> help(Expr.could_extract_minus_sign)
Help on method could_extract_minus_sign in module sympy.core.expr:

could_extract_minus_sign(self) unbound sympy.core.expr.Expr method
    Canonical way to choose an element in the set {e, -e} where
    e is any expression. If the canonical element is e, we have
    e.could_extract_minus_sign() == True, else  <--------- that True should be False
    e.could_extract_minus_sign() == False.  <------ that False should be True
    
    For any expression, the set ``{e.could_extract_minus_sign(), <---change "any expression" to "most expresions"
    (-e).could_extract_minus_sign()}`` must be ``{True, False}``. <--- change "must" to "will"
```
Add the sentence, "If e == -e then e.could_extract_minus_sign() will be False."

asmeurer commented:
The technically correct statement would be "if e.could_extract_minus_sign() is True, then (-e).could_extract_minus_sign() will be False".

The reverse implication won't hold if e == -e. But the way the function should be used is to negate if it the result is True, e.g., in an even function:

```
if arg.could_extract_minus_sign():
    arg = -arg
```

smichr commented:
How about this:

```python
        """Return True if self is not in a canonical form with respect
        to its sign.

        For most expressions, e, there will be a difference in e and -e.
        When there is, True will be returned for one and False for the
        other; False will be returned if there is no difference.

        Examples
        ========

        >>> from sympy.abc import x, y
        >>> e = x - y
        >>> {i.could_extract_minus_sign() for i in (e, -e)}
        {False, True}

        """
        negative_self = -self
        if self == negative_self:
            return False  # e.g. zoo*x == -zoo*x
        self_has_minus = (self.extract_multiplicatively(-1) is not None)
        negative_self_has_minus = (
            (negative_self).extract_multiplicatively(-1) is not None)
        if self_has_minus != negative_self_has_minus:
        ...
```

Dirivian commented:
Yeah, that is much more understandable.

One of the tests in `tests_expr.py` failed. Probably one of the two lines l added.

`    assert (zoo*(x - 1)).could_extract_minus_sign() is False`
`    assert (-(zoo*(1 - x ))).could_extract_minus_sign() is False`

I think that's because `could_extract_minus_sign` returns a method object. For example,
`(1-x).could_extract_minus_sign`
returns
` <bound method Expr.could_extract_minus_sign of -x + 1>`

Any idea why ?

smichr commented:
> Any idea why

You required that there be no args, but things like `zoo*x` have args:

```
if not self.args and self == -self:
```

smichr commented:
(Note the difference between what you wrote and what my new code proposes.)

Dirivian commented:
Ah, my bad.

Should we account for the case when the expression is 0 since 0=-0 ?

smichr commented:
> Should we account for the case when the expression is 0

I'm not sure I know what you mean by "account" -- it seems like this will just return False as it should, right? And we don't have to worry about symbolic zeros since there will be a canonical symbolic expression for which -1 cannot be extracted. e.g. 

```
>>> eq=x*(x - 1) - x**2 +x
>>> eq.expand()
0
>>> eq.could_extract_minus_sign()
False
>>> (-eq).could_extract_minus_sign()
True
```

The routine is giving an answer for the expression, not the underlying value.

smichr commented:
Hmm...I wonder if you mean that we could simply use `S.Zero.could_extract_minus_sign() is False` as the test. 

@asmeurer often writes little classes for tests like these, creating a hypothetical object that has the desired behavior and maybe that is the best way to test this otherwise one might think that only 0 (and possibly zoo) need to be special cased. Maybe

```
class sign_invariant(Function):  # 
    nargs = 1
    def __neg__(self):
        return self
foo = assert sign_invariant(x)
assert foo == -foo
assert foo.could_exract_minus_sign() is False
```

smichr commented:
Please add a note in the opening paragraph like was done in #13117 so the issue is closed when this is committed *and* so a reference to this PR is more clearly given on the issue page.

Dirivian commented:
I was thinking of a case in which an expression might be multiplied by/ ended up being turned to zero (the exact value not the underlying one) and whether something like that would be passed to `could_extract_minus_sign`. However, in that case, it's more of the programmer's decision and `could_extract_minus_sign` behaves as it should.

Added the note referencing the issue.

smichr commented:
Would you be willing to replace the zoo test with the snippet (sans #) that I gave above? That way, if the negation is ever modified to avoid putting the sign on a sign invariant factor the test will still work (whereas the current test may fail).

smichr commented:
The code should read

```python
class sign_invariant(Function):
    nargs = 1
    def __neg__(self):
        return self
foo = sign_invariant(x)
assert foo == -foo
assert foo.could_exract_minus_sign() is False
```

smichr commented:
Sorry -- you have to subclass from `Function, Expr`

smichr commented:
OK, how about `git reset head~7` then change `foo.could_exract_minus_sign() is False` to `foo.could_extract_minus_sign() is False` (note the t after the x) and make a single commit and this will be done.

smichr commented:
(You could add a space to make the `sign_invariant` definition read `class sign_invariant(Function, Expr)` while you are at it, too.)

Dirivian committed:
single commit

smichr committed:
could_extract_minus_sign is sign invariant aware

docstring and function have been modified to indicate
that for any foo where foo = -foo, could_extract_minus_sign
will return False. 0 and a Mul with a factor of zoo are two
examples of things that behave this way.

Negation of a Mul could be modified so that the first Add factor
could take the negation, but this is a separate issue. The added
test will continue to pass even if Mul is changed that way.

smichr commented:
I don't know why the error is being raised.

smichr commented:
Maybe you could rebase over master then repush using `git push -f master`.

asmeurer commented:
Here's the error. It's in the lines above. I guess this check will also be the first to warn about a syntax error. 

```
*** Sorry: IndentationError: unindent does not match any outer indentation level (expr.py, line 2193)
```

And unfortunately compileall has a ton of output that isn't helpful when there's an error. I wish we could find a better way.

asmeurer commented on sympy/core/expr.py:
```diff
@@ -2174,20 +2174,25 @@ def extract_additively(self, c):
-           e is any expression. If the canonical element is e, we have
-           e.could_extract_minus_sign() == True, else
-           e.could_extract_minus_sign() == False.
+         """Return True if self is not in a canonical form with respect
```
This line is indented too far.

Dirivian committed:
indentation fixed

smichr commented:
Thanks, all.

asmeurer commented:
Thanks for the fix. I've also found a related issue https://github.com/sympy/sympy/issues/13134. I don't know if it will be as straightforward to fix.
